core/trino-main/src/main/java/io/trino/server/protocol/Query.java

```diff
@@ -417,6 +417,8 @@ private synchronized QueryResults getNextResult(long token, UriInfo uriInfo, Dat
         QueryInfo queryInfo = queryManager.getFullQueryInfo(queryId);
         queryManager.recordHeartbeat(queryId);
 
+        closeExchangeClientIfNecessary(queryInfo);
```

Comment:
> Ensure cached query results are discarded
>
> Send next token one more time to make the client acknowledge results to
discard cached query results

I don't understand the commit message. the `GET next results` should be idempotent, and this sounds like it no longer is?

Reply:
Could you please elaborate more? From what I understand it isn't. The `Query#getNextResult` method calls `removePagesFromExchange` that transfers a next batch of pages from the `ExchangeClient` buffer and places it into it's own temporary buffer. When testing for memory leaks I found out that the `Query#lastResult` doesn't always get cleaned as the client doesn't try to acknowladge the last request (it doesn't come back as the nextUrl is not set)